zerver/tests/test_realm.py

```diff
@@ -620,6 +622,69 @@ def test_deactivate_realm_by_non_owner(self) -> None:
         realm = get_realm("zulip")
         self.assertFalse(realm.deactivated)
 
+    def test_delete_deactivated_realms(self) -> None:
+        lear = get_realm("lear")
+        zulip = get_realm("zulip")
+        zulipinternal = get_realm("zulipinternal")
+        zephyr = get_realm("zephyr")
+
+        all_realms_list = [realm.string_id for realm in Realm.objects.all()]
+        expected_realms_list = ["zulipinternal", "zephyr", "lear", "zulip"]
+        self.assertEqual(sorted(expected_realms_list), sorted(all_realms_list))
+
+        do_deactivate_realm(lear, acting_user=None, realm_deletion_in_minutes=0)
+        self.assertTrue(lear.deactivated)
+
+        all_realms_list = [realm.string_id for realm in Realm.objects.all()]
+        expected_realms_list = ["zulipinternal", "zephyr", "zulip"]
+        self.assertEqual(sorted(expected_realms_list), sorted(all_realms_list))
+
+        do_deactivate_realm(zulipinternal, acting_user=None, realm_deletion_in_minutes=10)
+        do_deactivate_realm(zulip, acting_user=None, realm_deletion_in_minutes=None)
+
+        self.assertTrue(zulipinternal.deactivated)
+        self.assertTrue(zulip.deactivated)
+
+        # zephyr realm is set to delete after 6 minutes
+        current_time = timezone_now()
+        with mock.patch(
+            "zerver.actions.realm_settings.timezone_now",
+            return_value=current_time + timedelta(minutes=3),
+        ):
```

Comment:
You should switch this to the `time_machine` library.

But it's not clear to me why you jump forward 3 minutes to ask to be deleted in 3 minutes, then jump back to now to check if it's _currently deleted_ (despite the deletion event being in the future at that time, which should be impossible), them jump forward 4 minutes, then back, then forward 6 minutes.

I think (with `time_machine`) this should be closer to:

- schedule the deletion in 3 minutes
- check that it's deactivated
- call delete_deactivated_realms, assert this doesn't do anything
- move forward 3 minutes
- call delete_deactivated_realms, assert `zephyr` is scrubbed
- move forward 7 more minutes
- call delete_deactivated_realms, assert `zulipinternal` is scrubbed

Reply:
@alexmv I need few suggestions:
- **For asserting that calling delete_deactivated_realms does not do anything**
 Should I assert that there is no change in `RealmAuditLog` entries after calling the function? Or there is anything better we can do.
- **For asserting that a realm is scrubbed**
I could not think anything better other than asserting that count of Messages, Attachments and CustomProfile Fields associated with realm is zero.